README.md
# Sharable eslint config with strict rules (limited formatting)

## Semver

- Patch version is incremented when existing rule is changed but eslint version is the same
- Minor version is incremented when config is updated to make use of new eslint rules or options. eslint dependency is also updated
- See [eslint semver policy](https://github.com/eslint/eslint#semantic-versioning-policy)

## Main features

- Rules are based on eslint version 8.21
- Most rules are configured in the strictest way possible with some exceptions
- Some rules are configured as warnings where a possible use case exists or when they are often broken during development
- Strict mode errors are checked for
- Most formatting rules are disabled
  - Some rules still conflict with prettier, notably no-mixed-operators
  - This config is meant to be used with [dprint](https://dprint.dev/) as formatter

## Code smells

- Maximum cyclomatic complexity is 3
- Maximum nesting depth is 3
- Maximum number of lines in a file is 300 (excluding empty lines & comments)
- Maximum number of lines in a function is 50 (including empty lines & comments)
- Maximum number of params in a function is 4
- Maximum number of statements in a function is 20

## Strictness

- Variable initialization is required
- Bitwise operators are forbidden
- eval is forbidden
- Short-hand type conversions are forbidden except for !! and +
- Magic numbers are forbidden (except for basic integers such as 1)
- Nested ternaries are allowed
- Shadowing variables is forbidden
- Explicit usage of undefined is forbidden (prefer null)
- var keyword is forbidden so its associated rules are disabled
- Constant assignment is preferred
- A number of regex strictness rules are applied

## Style rules

- camelCase is enforced for variable names
- Some non-specific variable names (e.g. res) are discouraged
- Unused variables must be prefixed with _
- Variable & property name length should not exceed 30 characters
- comments are uncapitalized
- Each variable declaration requires a separate keyword
- Function expressions are forbidden (prefer function declarations)
- Callbacks as arrow functions are preferred
- Object destructuring is preferred
- Imports are sorted
- Object keys are sorted if there are more than 10 of them (empty lines can be used to reset sorting)
- Blank lines are required between certain statement (e.g. before throw)


CHANGELOG.md
# v1.1.0

- Upgrade to eslint 8.21
- Update sort-keys rule


possibleProblems.js
module.exports = {
  possibleProblems: {
    'array-callback-return': [
      2,
      {
        allowImplicit: false,
        checkForEach: true,
      },
    ],
    'constructor-super': 2,
    'for-direction': 2,
    'getter-return': [
      2,
      {
        allowImplicit: false,
      },
    ],
    'no-async-promise-executor': 2,
    'no-await-in-loop': 1,
    'no-class-assign': 2,
    'no-compare-neg-zero': 2,
    'no-cond-assign': [2, 'except-parens'],
    'no-const-assign': 2,
    'no-constant-binary-expression': 2,
    'no-constant-condition': [
      2,
      {
        checkLoops: true,
      },
    ],
    'no-constructor-return': 2,
    'no-control-regex': 2,
    'no-debugger': 1,
    'no-dupe-args': 2,
    'no-dupe-class-members': 2, // useless in TS
    'no-dupe-else-if': 2,
    'no-dupe-keys': 2,
    'no-duplicate-case': 2,
    'no-duplicate-imports': [
      2,
      {
        includeExports: true,
      },
    ],
    'no-empty-character-class': 2,
    'no-empty-pattern': 2,
    'no-ex-assign': 2,
    'no-fallthrough': [
      2,
      {
        commentPattern: 'falls?\\s?through',
      },
    ],
    'no-func-assign': 2,
    'no-import-assign': 2,
    'no-inner-declarations': [2, 'functions'],
    'no-invalid-regexp': [
      2,
      {
        allowConstructorFlags: [],
      },
    ],
    'no-irregular-whitespace': [
      2,
      {
        skipStrings: false,
        skipComments: false,
        skipRegExps: false,
        skipTemplates: false,
      },
    ],
    'no-loss-of-precision': 2,
    'no-misleading-character-class': 2,
    'no-new-symbol': 2,
    'no-obj-calls': 2,
    'no-promise-executor-return': 2,
    'no-prototype-builtins': 2,
    'no-self-assign': [
      'error',
      {
        props: true,
      },
    ],
    'no-self-compare': 2,
    'no-setter-return': 2,
    'no-sparse-arrays': 2,
    'no-template-curly-in-string': 1,
    'no-this-before-super': 2,
    'no-undef': [
      2,
      {
        typeof: false,
      },
    ],
    'no-unexpected-multiline': 2,
    'no-unmodified-loop-condition': 2,
    'no-unreachable': 1,
    'no-unreachable-loop': [
      2,
      {
        ignore: [],
      },
    ],
    'no-unsafe-finally': 2,
    'no-unsafe-negation': [
      2,
      {
        enforceForOrderingRelations: true,
      },
    ],
    'no-unsafe-optional-chaining': [
      2,
      {
        disallowArithmeticOperators: true,
      },
    ],
    'no-unused-private-class-members': 1,
    'no-unused-vars': [
      1,
      {
        vars: 'all',
        varsIgnorePattern: '',
        args: 'all',
        argsIgnorePattern: '^_',
        ignoreRestSiblings: false,
        destructuredArrayIgnorePattern: '^_',
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '',
      },
    ],
    'no-use-before-define': [
      2,
      {
        functions: false,
        classes: true,
        variables: true,
        allowNamedExports: false,
      },
    ],
    'no-useless-backreference': 2,
    'require-atomic-updates': [
      2,
      {
        allowProperties: false,
      },
    ],
    'use-isnan': [
      2,
      {
        enforceForSwitchCase: true,
        enforceForIndexOf: true,
      },
    ],
    'valid-typeof': [
      2,
      {
        requireStringLiterals: true,
      },
    ],
  },
}


layoutAndFormatting.js
module.exports = {
  layoutAndFormatting: {
    'array-bracket-newline': 0,
    'array-bracket-spacing': 0,
    'array-element-newline': 0,
    'arrow-parens': [
      2,
      'as-needed',
      {
        requireForBlockBody: false,
      },
    ],
    'arrow-spacing': 0,
    'block-spacing': 0,
    'brace-style': 0,
    'comma-dangle': 0,
    'comma-spacing': 0,
    'comma-style': 0,
    'computed-property-spacing': 0,
    'dot-location': 0,
    'eol-last': 0,
    'func-call-spacing': 0,
    'function-call-argument-newline': 0,
    'function-paren-newline': 0,
    'generator-star-spacing': 0,
    'implicit-arrow-linebreak': 0,
    'indent': 0,
    'jsx-quotes': 0,
    'key-spacing': 0,
    'keyword-spacing': 0,
    'line-comment-position': 0,
    'linebreak-style': 0,
    'lines-around-comment': 0,
    'lines-between-class-members': [
      2,
      'always',
      {
        exceptAfterSingleLine: true,
      },
    ],
    'max-len': [
      2,
      {
        code: 120,
        tabWidth: 2,
        comments: 120,
        ignorePattern: '',
        ignoreComments: false,
        ignoreTrailingComments: false,
        ignoreUrls: false,
        ignoreStrings: false,
        ignoreTemplateLiterals: false,
        ignoreRegExpLiterals: true,
      },
    ],
    'max-statements-per-line': [
      2,
      {
        max: 1,
      },
    ],
    'multiline-ternary': 0,
    'new-parens': 0,
    'no-mixed-spaces-and-tabs': 0,
    'no-multi-spaces': 0,
    'no-multiple-empty-lines': 0,
    'no-tabs': [
      2,
      {
        allowIndentationTabs: false,
      },
    ],
    'no-trailing-spaces': 0,
    'no-whitespace-before-property': 0,
    'nonblock-statement-body-position': 0,
    'object-curly-newline': 0,
    'object-curly-spacing': 0,
    'object-property-newline': 0,
    'operator-linebreak': 0,
    'padded-blocks': 0,
    'padding-line-between-statements': [
      2,
      { blankLine: 'always', prev: 'block-like', next: '*' },
      { blankLine: 'always', prev: '*', next: 'class' },
      { blankLine: 'always', prev: 'class', next: '*' },
      { blankLine: 'always', prev: 'directive', next: '*' },
      { blankLine: 'never', prev: 'directive', next: 'directive' },
      { blankLine: 'always', prev: 'function', next: '*' },
      { blankLine: 'always', prev: '*', next: 'function' },
      { blankLine: 'always', prev: 'if', next: '*' },
      { blankLine: 'always', prev: 'import', next: '*' },
      { blankLine: 'any', prev: 'import', next: 'import' },
      { blankLine: 'always', prev: '*', next: 'throw' },
      { blankLine: 'always', prev: '*', next: 'cjs-export' },
      { blankLine: 'always', prev: 'cjs-import', next: '*' },
      { blankLine: 'any', prev: 'cjs-import', next: 'cjs-import' },
    ],
    'quotes': 0,
    'rest-spread-spacing': 0,
    'semi': 0,
    'semi-spacing': 0,
    'semi-style': 0,
    'space-before-blocks': 0,
    'space-before-function-paren': 0,
    'space-in-parens': 0,
    'space-infix-ops': 0,
    'space-unary-ops': 0,
    'switch-colon-spacing': 0,
    'template-curly-spacing': 0,
    'template-tag-spacing': 0,
    'unicode-bom': [
      2,
      'never',
    ],
    'wrap-iife': 0,
    'wrap-regex': 0,
    'yield-star-spacing': 0,
  },
}


suggestions.js
/* eslint-disable max-lines */
module.exports = {
  suggestions: {
    'accessor-pairs': [
      2,
      {
        setWithoutGet: true,
        getWithoutSet: false,
        enforceForClassMembers: true,
      },
    ],
    'arrow-body-style': [
      2,
      'as-needed',
      {
        requireReturnForObjectLiteral: false,
      },
    ],
    'block-scoped-var': 0,
    'camelcase': [
      2,
      {
        properties: 'always',
        ignoreDestructuring: false,
        ignoreImports: false,
        ignoreGlobals: false,
        allow: [],
      },
    ],
    'capitalized-comments': [
      1,
      'never',
      {
        line: {
          ignorePattern: '',
          ignoreInlineComments: false,
          ignoreConsecutiveComments: false,
        },
        block: {
          ignorePattern: '',
          ignoreInlineComments: false,
          ignoreConsecutiveComments: false,
        },
      },
    ],
    'class-methods-use-this': [
      1,
      {
        exceptMethods: [],
        enforceForClassFields: true,
      },
    ],
    'complexity': [
      1,
      {
        max: 3,
      },
    ],
    'consistent-return': [
      2,
      {
        treatUndefinedAsUnspecified: true,
      },
    ],
    'consistent-this': [2, 'that'],
    'curly': [2, 'multi-line', 'consistent'],
    'default-case': [
      2,
      {
        commentPattern: '^no default$',
      },
    ],
    'default-case-last': 2,
    'default-param-last': 2,
    'dot-notation': [
      2,
      {
        allowKeywords: true,
        allowPattern: '',
      },
    ],
    'eqeqeq': [2, 'smart'],
    'func-name-matching': [
      2,
      'always',
      {
        considerPropertyDescriptor: true,
        includeCommonJSModuleExports: false,
      },
    ],
    'func-names': [
      0,
      // as-needed doesn't work
      // requires names when assigning to object props
      // even though such functions are correctly displayed in error stacks
      'as-needed',
      {
        generators: 'as-needed',
      },
    ],
    'func-style': [
      2,
      'declaration',
      {
        allowArrowFunctions: false,
      },
    ],
    'grouped-accessor-pairs': [1, 'getBeforeSet'],
    'guard-for-in': 2,
    'id-denylist': [
      1,
      'res',
      'element',
      'prop',
      'property',
      'param',
      'parameter',
      'num',
      'number',
      'str',
      'string',
      'bool',
      'boolean',
      'obj',
      'object',
      'func',
      'foo',
      'bar',
      'baz',
    ],
    'id-length': [
      1,
      {
        // rule is too rigid
        // should have option to disable checks in array method callbacks
        // e.g. number.map(n => n * 2)
        min: 0,
        max: 30,
        properties: 'always',
        exceptions: ['k', 'v', 'el', 'ev', 'x', 'y'],
        exceptionPatterns: [],
      },
    ],
    'id-match': 0,
    'init-declarations': [2, 'always'],
    'max-classes-per-file': [
      2,
      {
        max: 1,
        ignoreExpressions: false,
      },
    ],
    'max-depth': [
      1,
      {
        max: 3,
      },
    ],
    'max-lines': [
      1,
      {
        max: 300,
        skipBlankLines: true,
        skipComments: true,
      },
    ],
    'max-lines-per-function': [
      1,
      {
        max: 50,
        skipBlankLines: false,
        skipComments: false,
        IIFEs: true,
      },
    ],
    'max-nested-callbacks': [
      2,
      {
        max: 3,
      },
    ],
    'max-params': [
      1,
      {
        max: 4,
      },
    ],
    'max-statements': [
      1,
      // can't be used in a object as a max property with another option
      // possibly a bug
      // eslint-disable-next-line no-magic-numbers
      20,
      {
        ignoreTopLevelFunctions: true,
      },
    ],
    'multiline-comment-style': [0, 'starred-block'],
    'new-cap': [
      2,
      {
        newIsCap: true,
        capIsNew: true,
        newIsCapExceptions: [],
        newIsCapExceptionPattern: '',
        capIsNewExceptions: [],
        capIsNewExceptionPattern: '',
        properties: true,
      },
    ],
    'no-alert': 1,
    'no-array-constructor': 2,
    'no-bitwise': [
      2,
      {
        allow: [],
        int32Hint: false,
      },
    ],
    'no-caller': 2,
    'no-case-declarations': 2,
    'no-confusing-arrow': [
      2,
      {
        allowParens: true,
        onlyOneSimpleParam: false,
      },
    ],
    'no-console': [
      0,
      {
        allow: [],
      },
    ],
    'no-continue': 0,
    'no-delete-var': 0, // useless in strict mode
    'no-div-regex': 2,
    'no-else-return': [
      2,
      {
        allowElseIf: false,
      },
    ],
    'no-empty': [
      1,
      {
        allowEmptyCatch: false,
      },
    ],
    'no-empty-function': [
      1,
      {
        // already covered by no-useless-constructor
        allow: ['constructors'],
      },
    ],
    'no-eq-null': 0,
    'no-eval': [
      2,
      {
        allowIndirect: true,
      },
    ],
    'no-extend-native': [
      2,
      {
        exceptions: [],
      },
    ],
    'no-extra-bind': 2,
    'no-extra-boolean-cast': [
      2,
      {
        enforceForLogicalOperands: true,
      },
    ],
    'no-extra-label': 2,
    'no-extra-semi': 2,
    'no-floating-decimal': 2,
    'no-global-assign': [
      2,
      {
        exceptions: [],
      },
    ],
    'no-implicit-coercion': [
      2,
      {
        /* eslint-disable id-denylist */
        boolean: true,
        number: true,
        string: true,
        /* eslint-enable id-denylist */
        disallowTemplateShorthand: true,
        allow: ['!!', '+'],
      },
    ],
    'no-implicit-globals': [
      2,
      {
        lexicalBindings: true,
      },
    ],
    'no-implied-eval': 2,
    'no-inline-comments': [
      0,
      {
        ignorePattern: '',
      },
    ],
    'no-invalid-this': [
      2,
      {
        capIsConstructor: true,
      },
    ],
    'no-iterator': 2,
    'no-label-var': 2,
    'no-labels': [
      0,
      {
        allowLoop: false,
        allowSwitch: false,
      },
    ],
    'no-lone-blocks': 2,
    'no-lonely-if': 2,
    'no-loop-func': 2,
    'no-magic-numbers': [
      2,
      {
        ignore: [-1, 0, 1, 2],
        ignoreArrayIndexes: false,
        ignoreDefaultValues: true,
        enforceConst: true,
        detectObjects: false,
      },
    ],
    'no-mixed-operators': [
      2,
      {
        groups: [
          ['+', '-', '*', '/', '%', '**', '?:'],
          ['&', '|', '^', '~', '<<', '>>', '>>>', '?:'],
          ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
          ['&&', '||', '??'],
          ['in', 'instanceof', '?:'],
        ],
        allowSamePrecedence: true,
      },
    ],
    'no-multi-assign': [
      2,
      {
        ignoreNonDeclaration: false,
      },
    ],
    'no-multi-str': 0,
    'no-negated-condition': 2,
    'no-nested-ternary': 0,
    'no-new': 2,
    'no-new-func': 2,
    'no-new-object': 2,
    'no-new-wrappers': 2,
    'no-nonoctal-decimal-escape': 2,
    'no-octal': 0, // useless in strict mode
    'no-octal-escape': 0, // useless in strict mode
    'no-param-reassign': [
      // the rule is too rigid
      // should have an option to disable checking for array methods callbacks
      // e.g. reduce, forEach
      0,
      {
        props: true,
        ignorePropertyModificationsFor: [],
        // eslint-disable-next-line id-length
        ignorePropertyModificationsForRegex: [],
      },
    ],
    'no-plusplus': [
      0,
      {
        allowForLoopAfterthoughts: true,
      },
    ],
    'no-proto': 2,
    'no-redeclare': [
      2,
      {
        builtinGlobals: true,
      },
    ],
    'no-regex-spaces': 2,
    'no-restricted-exports': [
      2,
      {
        restrictedNamedExports: [],
      },
    ],
    'no-restricted-globals': [0],
    'no-restricted-imports': [
      0,
      {
        paths: [],
        patterns: [],
      },
    ],
    'no-restricted-properties': [0],
    'no-restricted-syntax': [0],
    'no-return-assign': [2, 'except-parens'],
    'no-return-await': 0,
    'no-script-url': 0,
    'no-sequences': [
      2,
      {
        allowInParentheses: true,
      },
    ],
    'no-shadow': [
      2,
      {
        builtinGlobals: true,
        hoist: 'functions',
        ignoreOnInitialization: false,
        allow: [],
      },
    ],
    'no-shadow-restricted-names': 0,
    'no-ternary': 0,
    'no-throw-literal': 2,
    'no-undef-init': 2,
    'no-undefined': 0,
    'no-underscore-dangle': [
      2,
      {
        allow: [],
        allowAfterThis: false,
        allowAfterSuper: false,
        allowAfterThisConstructor: false,
        enforceInMethodNames: false,
        enforceInClassFields: false,
        allowFunctionParams: true,
      },
    ],
    'no-unneeded-ternary': [
      2,
      {
        defaultAssignment: false,
      },
    ],
    'no-unused-expressions': [
      2,
      {
        allowShortCircuit: true,
        allowTernary: false,
        allowTaggedTemplates: false,
        enforceForJSX: true,
      },
    ],
    'no-unused-labels': 2,
    'no-useless-call': 2,
    'no-useless-catch': 2,
    'no-useless-computed-key': [
      2,
      {
        enforceForClassMembers: true,
      },
    ],
    'no-useless-concat': 2,
    'no-useless-constructor': 2,
    'no-useless-escape': 2,
    'no-useless-rename': [
      2,
      {
        ignoreImport: false,
        ignoreExport: false,
        ignoreDestructuring: false,
      },
    ],
    'no-useless-return': 2,
    'no-var': 2,
    'no-void': [
      2,
      {
        allowAsStatement: true,
      },
    ],
    'no-warning-comments': [
      0,
      {
        terms: ['todo', 'fixme', 'xxx'],
        location: 'anywhere',
      },
    ],
    'no-with': 0, // useless in strict mode
    'object-shorthand': [
      2,
      'always',
      {
        avoidQuotes: true,
        ignoreConstructors: false,
        avoidExplicitReturnArrows: true,
      },
    ],
    'one-var': [
      2,
      {
        var: 'never',
        let: 'never',
        const: 'never',
        separateRequires: true, // possibly doesn't work
      },
    ],
    'one-var-declaration-per-line': [
      0, // useless with one-var set no never
      'initializations',
    ],
    'operator-assignment': [2, 'always'],
    'prefer-arrow-callback': [
      2,
      {
        allowNamedFunctions: false,
        allowUnboundThis: false,
      },
    ],
    'prefer-const': [
      2,
      {
        destructuring: 'all',
        ignoreReadBeforeAssign: false,
      },
    ],
    'prefer-destructuring': [
      2,
      {
        /* eslint-disable id-denylist */
        VariableDeclarator: {
          array: true,
          object: true,
        },
        AssignmentExpression: {
          array: true,
          object: true,
        },
        /* eslint-enable id-denylist */
      },
      {
        enforceForRenamedProperties: false,
      },
    ],
    'prefer-exponentiation-operator': 2,
    'prefer-named-capture-group': 2,
    'prefer-numeric-literals': 2,
    'prefer-object-has-own': 2,
    'prefer-object-spread': 2,
    'prefer-promise-reject-errors': [
      2,
      {
        allowEmptyReject: false,
      },
    ],
    'prefer-regex-literals': [
      2,
      {
        disallowRedundantWrapping: true,
      },
    ],
    'prefer-rest-params': 2,
    'prefer-spread': 2,
    'prefer-template': 0,
    'quote-props': [
      2,
      'consistent-as-needed',
      // keyword object prop can't be used
      // possibly a bug
    ],
    'radix': [2, 'as-needed'],
    'require-await': 0,
    'require-unicode-regexp': 2,
    'require-yield': 2,
    'sort-imports': [
      2,
      {
        ignoreCase: true,
        ignoreDeclarationSort: false,
        ignoreMemberSort: false,
        memberSyntaxSortOrder: ['none', 'all', 'multiple', 'single'],
        allowSeparatedGroups: true,
      },
    ],
    'sort-keys': [
      2,
      'asc',
      {
        caseSensitive: false,
        minKeys: 11,
        natural: true,
      },
    ],
    'sort-vars': [
      2,
      {
        ignoreCase: true,
      },
    ],
    'spaced-comment': [
      2,
      'always',
      {
        line: {
          markers: ['/'],
          exceptions: [],
        },
        block: {
          markers: [],
          exceptions: [],
          balanced: true,
        },
      },
    ],
    'strict': [
      1,
      'never', // useless because of impliedStrict in parserOptions
    ],
    'symbol-description': 2,
    'vars-on-top': 0,
    'yoda': [
      2,
      'never',
      {
        exceptRange: true,
        onlyEquality: false,
      },
    ],
  },
}
